File: src/main/java/com/coalesce/commands/executors/Help.java
package com.coalesce.commands.executors;

import com.coalesce.commands.Command;
import com.coalesce.commands.CommandExecutor;
import com.coalesce.commands.CommandMap;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.MessageBuilder;
import net.dv8tion.jda.core.entities.Message;
import net.dv8tion.jda.core.entities.MessageChannel;

import java.awt.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Command(name = "help", description = "Show bot help", aliases = { "?", "h" })
public class Help extends CommandExecutor {
    
    @Override
    protected void execute(MessageChannel channel, Message message, String[] args) throws Exception {
        MessageBuilder mb = new MessageBuilder();
        mb.append("```\n");
        
        List<CommandMap.CommandEntry> entries = new ArrayList<>();
        
        commandMap.getEntries().values().forEach(e -> {
            if (!entries.contains(e)) {
                entries.add(e);
                mb.append(e.getMeta().name()).append(" ").append(e.getMeta().usage()).append(" ")
                        .append(e.getMeta().aliases().length == 0
                                ? ""
                                : Arrays.toString(e.getMeta().aliases()))
                        .append("\n");
            }
        });
        
        entries.clear();
        
        mb.append("```");
        
        channel.sendMessage(new EmbedBuilder()
                .setTitle("Help", null)
                .setColor(Color.GREEN)
                .setDescription(mb.build().getRawContent())
                .build()).queue();
    }
}

File: src/main/kotlin/com/coalesce/commands/executors/Help.kt
package com.coalesce.commands.executors

import com.coalesce.commands.Command
import com.coalesce.commands.CommandExecutor
import com.coalesce.commands.CommandType
import net.dv8tion.jda.core.EmbedBuilder
import net.dv8tion.jda.core.entities.Message
import net.dv8tion.jda.core.entities.MessageChannel
import java.awt.Color
import java.util.*
import java.util.concurrent.TimeUnit

@Command(name = "help", description = "Lists the commands", aliases = arrayOf("?", "h"), permission = "commands.help",
        globalCooldown = 35, type = CommandType.INFORMATION)
class Help : CommandExecutor() {
    override fun execute(channel: MessageChannel, message: Message, args: Array<String>) {
        val embedBuilder = EmbedBuilder().setTitle("Help", null).setColor(Color.GREEN)

        val map = mutableMapOf<CommandType, MutableSet<CommandExecutor>>().withDefault { mutableSetOf() }
        commandMap.entries.values.forEach {
            map[it.annotation.type] = map[it.annotation.type]?.apply { add(it.executor) } ?: mutableSetOf()
        }
        val out = mutableMapOf<CommandType, String>().withDefault { "None" }
        map.forEach { type, executors ->
            val builder = StringBuilder().append("```\n")
            executors.forEach {
                if (it.annotation.name.equals("xD", false)) {
                    builder.append("xD")
                } else {
                    builder.append(it.annotation.name.capitalize())
                }
                builder.append(' ').append(it.annotation.usage)
                if (!it.annotation.aliases.isEmpty()) {
                    builder.append(' ').append(Arrays.toString(it.annotation.aliases.flatMapTo(mutableSetOf<String>(), { setOf(it.capitalize()) }).toTypedArray()))
                }
                builder.append("\n")
            }
            out[type] = builder.append("```").toString().trim()
        }

        embedBuilder.addField("Fun", out[CommandType.FUN], true)
                .addField("Information", out[CommandType.INFORMATION], true)
                .addField("Administration", out[CommandType.ADMINISTRATION], true)
                .addField("Debug", out[CommandType.DEBUG], true)

        channel.sendMessage(embedBuilder.build()).queue { it.delete().queueAfter(25, TimeUnit.SECONDS) }
    }
}
